File: RPC_s/RPC2021/01/Geometric.py
mod = 1000000000

def fexp(a,b):
    ans=1
    while(b>0):
        if(b&1==1):
            ans=ans*a
        a=a*a
        b//=2
    return ans


t=int(input())
for T in range (t):
    l=list(map(int,input().split()))
    k=l[0]
    q=l[1]
    v=list(map(int,input().split()))
    for Q in range (q):
        a=int(input())
        if(a<k):
            print(v[a],end='\n',flush=False)
        else:
            num=v[a%k]
            exp=(a//k)+1
            print(fexp(num,exp)%mod,end='\n',flush=False)

File: RPC_s/RPC2021/01/Geometric.cpp
/// Write by Daniel Perez .PERAPRO
#include<bits/stdc++.h>

using namespace std;
#define fast_io ios_base::sync_with_stdio(0); cin.tie(0); cout.tie(0);
using vi=vector<int>;
using pii=pair<int,int>;
#define pb push_back
#define ff first
#define ss second
typedef long long ll;
typedef unsigned long long ull;
char el = '\n';

ostream& operator<<(ostream& os, const vector<int> &v){
    for(auto const &i: v){
        os<<i<<" ";
    }
    os<<endl;
    return os;
}

ll mod=1e9;

ll mul(ll a, ll b){
    return(__int128)a*b%mod;
}

ll ans;
ll fexp(ll a, ll b){
    for(;b;a=mul(a,a),b>>=1){
        if(b&1)
            ans=mul(ans,a);
    }
    return ans;
}

int main(){
	fast_io;
	int t;
	cin>>t;
	while(t--){
	    ll k,q;
	    cin>>k>>q;
	    vi v(k);
	    for(int i=0;i<k;i++) cin>>v[i];
	    //cout<<v;
	    while(q--){
	        ll a;cin>>a;
	        if(a<k) cout<<v[a]<<el;
	        else{
	            ll num=v[a%k];
	            ll ep=a/k+1;
	            if(num<0 && (ep&1)){
	                while(num<0)num+=mod;
	                ans=num;
	                cout<<fexp(abs(v[a%k]),ep-1)<<el;
	            }else{
	                ans=1;
	                cout<<fexp(abs(num),ep)<<el;
	            }
	        }
	    }
	}
}
